services/src/api/mod.rs

```diff
@@ -1 +1,699 @@
+use crate::contexts::SessionId;
+use crate::handlers::ErrorResponse;
+use actix_web::dev::ServiceResponse;
+use actix_web::http::{header, Method};
+use actix_web::test::TestRequest;
+use actix_web_httpauth::headers::authorization::Bearer;
+use std::collections::{BTreeMap, HashMap};
+use std::future::Future;
+use utoipa::openapi::path::{Parameter, ParameterIn};
+use utoipa::openapi::schema::AdditionalProperties;
+use utoipa::openapi::{
+    KnownFormat, OpenApi, PathItemType, Ref, RefOr, Schema, SchemaFormat, SchemaType,
+};
+use uuid::Uuid;
+
 pub mod model;
+
+fn get_schema_name_from_ref(reference: &Ref) -> &str {
+    const SCHEMA_REF_PREFIX_LEN: usize = "#/components/schemas/".len();
+    &reference.ref_location[SCHEMA_REF_PREFIX_LEN..]
+}
+
+fn throw_if_invalid_ref(reference: &Ref, schemas: &BTreeMap<String, RefOr<Schema>>) {
+    let schema_name = get_schema_name_from_ref(reference);
+    assert!(
+        schemas.contains_key(schema_name),
+        "Referenced the unknown schema `{schema_name}`"
+    );
+}
+
+fn can_resolve_schema(schema: RefOr<Schema>, schemas: &BTreeMap<String, RefOr<Schema>>) {
+    match schema {
+        RefOr::Ref(reference) => {
+            throw_if_invalid_ref(&reference, schemas);
+        }
+        RefOr::T(concrete) => match concrete {
+            Schema::Array(arr) => {
+                can_resolve_schema(*arr.items, schemas);
+            }
+            Schema::Object(obj) => {
+                for property in obj.properties.into_values() {
+                    can_resolve_schema(property, schemas);
+                }
+                if let Some(additional_properties) = obj.additional_properties {
+                    if let AdditionalProperties::RefOr(properties_schema) = *additional_properties {
+                        can_resolve_schema(properties_schema, schemas);
+                    }
+                }
+            }
+            Schema::OneOf(oo) => {
+                for item in oo.items {
+                    can_resolve_schema(item, schemas);
+                }
+            }
+            Schema::AllOf(ao) => {
+                for item in ao.items {
+                    can_resolve_schema(item, schemas);
+                }
+            }
+            _ => panic!("Unknown schema type"),
+        },
+    }
+}
+
+pub fn can_resolve_api(api: OpenApi) {
```

Comment:
Could you add some documentation on what this does in some words?

Reply:
[b0b7de2](https://github.com/geo-engine/geoengine/pull/760/commits/b0b7de2d20575d232ef0cbcb4b139b4600770743)